### ISSUE TITLE
"listener.cattle.io/cn" system adds artibtrary (unauthed) requests to annotation, rejects SNI hostnames > 41 characters 

### ISSUE THREAD

**Issue by erulabs** [k3s-io/k3s#4542]
**Environmental Info:**
K3s Version: 
k3s version v1.22.3+k3s1 (61a2aab2)
go version go1.16.8

Node(s) CPU architecture, OS, and Version: 
Linux pasadena 5.11.0-40-generic #44-Ubuntu SMP Wed Oct 20 16:16:42 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

Cluster Configuration:
1 server

**Describe the bug:**
Crafting a specific request to the k3s api causes some system to add "listener.cattle.io/cn-HOSTNAME" annotations. This seems questionable from a performance and security perspective, but also means that because annotations are limited to 63 characters, and `63 - 'listener.cattle.io/cn-'.length === 41`, any hostname longer than 41 characters will crash the k3s server and trigger a TLS error on the request.

**Steps To Reproduce:**
- Installed K3s via script, no options provided (completely default)
- Access k3s api with curl like:
```bash
$ curl https://CURL-SAYS-HELLO -k --resolve CURL-SAYS-HELLO:443:10.43.0.1
```
You get a 401 "Unauthorized", which is expected, however, k3s has patched the `k3s-serving` secret with an annotations:

```bash
$ kubectl get secrets -n kube-system k3s-serving -o yaml | fgrep HELLO
    listener.cattle.io/cn-CURL-SAYS-HELLO: CURL-SAYS-HELLO
```

This seems... Not great. At very least I can easily pollute this list by just curling the endpoint with different SNI headers.

Next, with a long enough cluster name:

```bash
$ curl -k https://CURL-SAYS-HELLO.this-is-a-very-long-domain-name-man-its-crazy.com --resolve CURL-SAYS-HELLO.this-is-a-very-long-domain-name-man-its-crazy.com:443:10.43.0.1
curl: (35) error:14094438:SSL routines:ssl3_read_bytes:tlsv1 alert internal error
```

This is because `k3s server` has errored trying to set that annotation:
```
Cluster-Http-Server 2021/11/18 22:27:11 http: TLS handshake error from 10.42.0.11:33676: Secret \"k3s-serving\" is invalid: metadata.annotations: Invalid value: \"listener.cattle.io/cn-CURL-SAYS-HELLO.this-is-a-very-long-domain-name-man-its-crazy.com\": name part must be no more than 63 characters
```

Ack! 😢 

**Expected behavior:**
I _suppose_ i see why,  if it's intentional, you might want to auto-add TLS SANs. It's actually maybe even really neat. But, storing them as annotation -keys- seems like a bad idea, adding them to unauthenticated requests seems bad too? At the very least, they should be stored in a way that allows for full-length hostnames. An example is when I use `raspberrypi.erulabs.kubesail-gateway.default.svc.cluster.local`.

**Actual behavior:**
TLS responses are broken when k3s is addressed using a very long (>41 character hostname)

**Backporting**
- [ ] Needs backporting to older releases

**Comment by brandond** [comment]
Yes, the on-demand addition of CNs is intentional, to allow for dynamic environments where the full set of hostnames and IP addresses might not be known at the time the service is started, or where they need to be added without action on the part of the administrator. We have considered allowing the ability to restrict this, but it hasn't been a priority.

The long hostname issue is addressed by a pending PR at https://github.com/rancher/dynamiclistener/pull/49

**Comment by erulabs** [comment]
Ah, it's hashed, perfect. Yes, a feature to disable it might be nice, but don't consider this a feature request 🙉 

https://github.com/rancher/dynamiclistener/pull/49 looks perfect - Thank you! Feel free to close this issue.

**Comment by qrzeller** [comment]
Thank you so much @erulabs.
I got an error message with **_kubectl_** showing a bunch of weird domain name. I thought i got hacked or at least an attempt. I lost a lot of time looking for the issue but this post somehow saved me a lot of time.
THANKS !
